### PR TITLE
fabtests/getinfo_test, prov/psm2, rxm, rxd, shm, udp, verbs, sockets, efa: test capability sets

### PR DESCRIPTION
**Pull request by aingerson**
- Fix some getinfo_test bugs
- Fix providers failing after bug fixes
- Add capability set testing to the getinfo_test
- Fix all providers that report incorrect capability sets

The man pages were recently updated to clarify which capabilities should be reported in the tx_attr->caps and rx_attr->caps. 

The ordering combinations and mr_mode combinations tests were refactored to be used generally for other combination tests and the domain_attr test was expanded to use this combination test and include all combination of domain_attr caps

A new test was added to test capability set combinations. This test iterates through all combinations of primary capabilities and makes sure that if primary capabilities were requested:
- only the requested capabilities are returned
- the tx and rx attr caps only return valid tx/rx capabilities
- the capabilities in the tx and rx match what is returned in the info->caps

### CODE REVIEW

File: fabtests/unit/getinfo_test.c [hunk 1]
```diff
@@ -110,91 +112,212 @@ static int invalid_dom(struct fi_info *hints)
...
+#define PRIMARY_TX_CAPS	(FI_MSG | FI_RMA | FI_TAGGED | FI_ATOMIC | \
+			 FI_MULTICAST | FI_NAMED_RX_CTX | FI_HMEM)
+#define PRIMARY_RX_CAPS (FI_MSG | FI_RMA | FI_TAGGED | FI_ATOMIC | \
+			 FI_DIRECTED_RECV | FI_VARIABLE_MSG | \
+			 FI_HMEM)
+
+#define PRIMARY_CAPS (PRIMARY_TX_CAPS | PRIMARY_RX_CAPS)
+#define DOMAIN_CAPS (FI_LOCAL_COMM | FI_REMOTE_COMM | FI_SHARED_AV)
+#define SEC_TX_CAPS (FI_TRIGGER | FI_FENCE | FI_RMA_PMEM)
+#define SEC_RX_CAPS (FI_RMA_PMEM | FI_VARIABLE_MSG | FI_SOURCE | \
```
[the remark this answers is not in the document]
Comment (aingerson) [reply fabtests/unit/getinfo_test.c]: Good catch, thank you!

**Comment by aingerson** [comment]
@rajachan  EFA was failing this test so I've split of the capability sets to hopefully pass. I can't compile or run EFA so I'm hoping this will only need one run but probably will need another round... Please take a look to make sure I'm not messing things up

**Comment by rajachan** [comment]
> @rajachan EFA was failing this test so I've split of the capability sets to hopefully pass. I can't compile or run EFA so I'm hoping this will only need one run but probably will need another round... Please take a look to make sure I'm not messing things up

Looks like the getinfo test is still failing with these changes:
https://libfabric-ci.aws.a2z.com/job/single-node/52/

We will take a look.

**Comment by aingerson** [comment]
@j-xiong Updated psm2 patch to remove FI_REMOTE_CQ_DATA from info->caps as well

**Comment by j-xiong** [comment]
> 
> 
> @j-xiong Updated psm2 patch to remove FI_REMOTE_CQ_DATA from info->caps as well

What is the reason behind this change?  Is this cap no longer needed?

**Comment by aingerson** [comment]
@j-xiong  FI_REMOTE_CQ_DATA is not a capability but rather an op_flag. The applications can determine or request cq data support by setting the domain_attr->cq_data_size field but not as a capability.

**Comment by j-xiong** [comment]
@aingerson This used to be a cap bit. I did a search and found it was removed a few years ago (commit cb1cf4f9550173d706a73267c4f5f3fc96513255). Somehow the psm2 provider didn't catch the change. Thanks for the update.

**Comment by rajachan** [comment]
https://github.com/ofiwg/libfabric/pull/5641, once merged, should resolve the EFA fi_getinfo_test failure.

**Comment by rajachan** [comment]
@aingerson try rebasing this PR on master again, the EFA fix got merged.

**Comment by aingerson** [comment]
@rajachan looks good now, thanks for fixing!

**Pull request merged**